nomenklatura/store/sql.py

```diff
@@ -0,0 +1,167 @@
+from typing import Any, Generator, List, Optional, Set, Tuple
+
+from followthemoney.property import Property
+from sqlalchemy import Table, create_engine, delete, select
+from sqlalchemy.sql.selectable import Select
+
+from nomenklatura.dataset import DS
+from nomenklatura.db import (
+    DB_URL,
+    POOL_SIZE,
+    get_metadata,
+    get_statement_table,
+    get_upsert_func,
+)
+from nomenklatura.entity import CE
+from nomenklatura.resolver import Resolver
+from nomenklatura.statement import Statement
+from nomenklatura.statement.serialize import pack_sql_statement
+from nomenklatura.store import Store, View, Writer
+
+
+class SqlStore(Store[DS, CE]):
+    def __init__(
+        self,
+        dataset: DS,
+        resolver: Resolver[CE],
+        uri: str = DB_URL,
+        **engine_kwargs: Any,
+    ):
+        super().__init__(dataset, resolver)
+        engine_kwargs["pool_size"] = engine_kwargs.pop("pool_size", POOL_SIZE)
+        self.metadata = get_metadata()
+        self.engine = create_engine(uri, **engine_kwargs)
+        self.table = get_statement_table()
+        self.metadata.create_all(self.engine, tables=[self.table], checkfirst=True)
+
+    def writer(self) -> Writer[DS, CE]:
+        return SqlWriter(self)
+
+    def view(self, scope: DS, external: bool = False) -> View[DS, CE]:
+        return SqlView(self, scope, external=external)
+
+    def _iterate_stmts(self, q: Select) -> Generator[Statement, None, None]:
+        with self.engine.connect() as conn:
+            conn = conn.execution_options(stream_results=True)
+            cursor = conn.execute(q)
+            while rows := cursor.fetchmany(10_000):
+                for row in rows:
+                    yield Statement.from_db_row(row)
+
+    def _iterate(self, q: Select) -> Generator[CE, None, None]:
+        current_id = None
+        current_stmts: list[Statement] = []
+        for stmt in self._iterate_stmts(q):
+            entity_id = stmt.entity_id
+            if current_id is None:
+                current_id = entity_id
+            if current_id != entity_id:
+                proxy = self.assemble(current_stmts)
+                if proxy is not None:
+                    yield proxy
+                current_id = entity_id
+                current_stmts = []
+            current_stmts.append(stmt)
+        if len(current_stmts):
+            proxy = self.assemble(current_stmts)
+            if proxy is not None:
+                yield proxy
+
+
+class SqlWriter(Writer[DS, CE]):
+    BATCH_STATEMENTS = 10_000
+
+    def __init__(self, store: SqlStore[DS, CE]):
+        self.store: SqlStore[DS, CE] = store
+        self.batch: Optional[Set[Statement]] = None
+        self.batch_size = 0
```

Comment:
Do we need `batch_size` if we have a `batch` array? I assume `len(batch)` is more precise.

Reply:
https://github.com/opensanctions/nomenklatura/pull/125/commits/3d25cb41a466a94325f3ff9f4f6ff1842b718bf4

(this originally came from my believe that `int > int` is more performant than `len(foo) > int` but gosh, this would never be the bottleneck in the overall sql store implementation) :joy: :see_no_evil: